tools/mo/openvino/tools/mo/main.py

```diff
@@ -310,24 +319,45 @@ def raise_ie_not_found():
         from openvino.tools.mo.front.onnx.register_custom_ops import get_front_classes
         import_extensions.load_dirs(argv.framework, extensions, get_front_classes)
 
-    return argv
 
+def check_fallback(argv : argparse.Namespace):
+    fallback_reasons = {}
+    fallback_reasons['extensions'] = \
+        lambda argv : hasattr(argv, 'extensions') and argv.extensions is not None and len(argv.extensions) > 0 \
+            and argv.extensions != import_extensions.default_path() # extensions arg has default value
+    fallback_reasons['transformations_config'] = \
+        lambda argv: hasattr(argv, 'transformations_config') and argv.transformations_config is not None and len(argv.transformations_config) > 0
+
+    reasons = [reason for reason, is_applicable in fallback_reasons.items() if is_applicable(argv)]
+    return reasons
 
-def prepare_ir(argv):
-    argv = arguments_post_parsing(argv)
 
+def prepare_ir(argv : argparse.Namespace):
+    argv = arguments_post_parsing(argv)
     t = tm.Telemetry()
     graph = None
     ngraph_function = None
     moc_front_end, available_moc_front_ends = get_moc_frontends(argv)
-
-    if moc_front_end:
+    fallback_reasons = check_fallback(argv)
+    if moc_front_end and len(fallback_reasons) == 0:
         t.send_event("mo", "conversion_method", moc_front_end.get_name() + "_frontend")
         moc_front_end.add_extension(TelemetryExtension("mo", t.send_event, t.send_error, t.send_stack_trace))
-        ngraph_function = moc_pipeline(argv, moc_front_end)
-    else:
-        t.send_event("mo", "conversion_method", "mo_legacy")
-        graph = unified_pipeline(argv)
+        try:
+            ngraph_function = moc_pipeline(argv, moc_front_end)
+            return graph, ngraph_function
+        except Exception as e:
+            fallback_reasons.append(f"frontend failure with exception: {e}")
+    if len(fallback_reasons) > 0:
+        argv.use_legacy_frontend = True
+        load_extensions(argv)
+        reasons_message = ", ".join(fallback_reasons)
+        log.warning("The IR preparation was executed by the legacy MO path. "
+                    "This is a fallback scenario applicable only for some specific cases. "
+                   f"The detailed reason why fallback was executed: {reasons_message}. "
+                    "You can specify --use_legacy_frontend flag to force using the legacy MO path to avoid additional checks.")
```

Comment:
Also, I have question how we are going to test new frontend features (that cover the current limitations) in our validation, if it automatically fallbacks legacy frontend. For example, once we start to support extensions and transformation_config files how will we track regressions?

Reply:
The only models from the current e2e validation which I've found as affected by fallback are maskrcnn and fasterrcnn (`--transformations_config` used).

They are tested now: 
- in one pipeline: without any flag (so legacy is the default path) 
- and with `--use_new_frontend` explicitly (so on new frontend without fallback - after applying https://github.com/openvinotoolkit/openvino/pull/9034#discussion_r775787984)

If the new frontend will be the default path, we can ask the validation team to run maskrcnn and fasterrcnn (without cmd flag) with expected pass result and additionally with `--use_new_frontend` (without fallback) with expected XFAIL.

@rkazants What do you think? Should I create a ticket for it?